### DISCUSSION TITLE
How to enforce row level security on the server side?

### DISCUSSION THREAD

**Discussion by alexreyes** [supabase/supabase#7000] [Questions] [answered]
I'm doing all the DB queries for my supabase project in the server side. However, since I use a connection string, the row level security policies aren't being applied to requests.

I want to enforce the row level security policies I've made for users making requests to my server. 

Essentially, I want to "[make requests "as" a particular user on the server](https://github.com/supabase/supabase/discussions/1094)". From the discussion linked in the prior sentence I found out about `supabase.auth.setAuth(jwt);`, but it seems to me like it's inefficient to create a new Supabase client for each request. 

## **Is there a better way to apply row level security rules to requests being made on behalf of a user in the server?**

Related discussions (I've read through all of these): 
- [Is it possible to make requests "as" a particular user on the server? #1094](https://github.com/supabase/supabase/discussions/1094)
- [Authentication in a serverless function](https://github.com/supabase/supabase/discussions/1067)
- [Auth row level security (supabase docs)](https://supabase.com/docs/learn/auth-deep-dive/auth-row-level-security)
- [Client & server side method for validating user auth token #147](https://github.com/supabase/supabase/issues/147)

**Comment by olirice** [comment, marked as answer]
> since I use a connection string, the row level security policies aren't being applied to requests.

On your backend server are you using supabase-js to make database queries, or using the connection string to query the database using SQL?

If using supabase-js, the solution in the links you referenced using `createClient` is the recommended answer.

If you're making queries via SQL, we need to take on the identity of the requesting user the same way that postgrest does to enable the RLS helpers,  `auth.uid()` and `auth.role()`, to return the correct values

To do that, within a transaction you can
```sql
begin; -- start transaction

    -- provide the user's JWT data to postgres config (read by auth.uid() and auth.role()) 
    set_config('request.jwt.claims', 
        -- provide JWT claims info here
        jsonb_build_object(
            'role', 'authenticated',
            'sub', '<user_uuid>'::text
        )::text,
        true
    );
    -- set the postgres role to use for executing the query
    set_config('role', 'authenticated', true);

    -- YOUR QUERY HERE
end;
```

and security + RLS policies will be applied according to the config provided

**Reply by awalias** [reply]
in addition to Oli's answer it's worth noting that it's actually pretty efficient to create a supabase client, since a client does not establish a connection with the database directly, it is merely a lightweight wrapper around a HTTP request (and is also therefore "connectionless")

to make it even more efficient (although I'm certain it will be negligible gains) you could create the postgrest-js client directly, and set the headers manually (this is what `.setAuth(jwt)` does for you)

or you could skip the client libs and just make the HTTP request directly, you just need to make a request like:

```bash
curl 'https://PROJECT_REF.supabase.co/rest/v1/countries?select=id' \
-H "apikey: SUPABASE_KEY" \
-H "Authorization: Bearer USER_JWT"
```
(note I plucked this code directly from the auto generated docs at `https://app.supabase.com/project/PROJECT_REF/api` and clicked `bash`)

**Reply by alexreyes** [reply]
Okay thanks for the responses! I ended up just creating a new supabase client each time a request is made, and I use `.setAuth(jwt)` to enforce the RLS. This was very helpful
